File: services/tms/internal/core/ports/services/invoice.go
package services

import (
	"context"

	"github.com/emoss08/trenova/internal/core/domain/invoice"
	"github.com/emoss08/trenova/internal/core/ports/repositories"
	"github.com/emoss08/trenova/pkg/pagination"
	"github.com/emoss08/trenova/shared/pulid"
)

type CreateInvoiceFromBillingQueueRequest struct {
	BillingQueueItemID pulid.ID
	TenantInfo         pagination.TenantInfo
}

type CreateInvoiceFromBillingQueueResult struct {
	Invoice  *invoice.Invoice
	AutoPost bool
}

type PostInvoiceRequest struct {
	InvoiceID   pulid.ID
	TenantInfo  pagination.TenantInfo
	TriggeredBy string
}

type CreateInvoiceFromShipmentsRequest struct {
	ShipmentIDs []pulid.ID
	TenantInfo  pagination.TenantInfo
}

type UpdateInvoiceDraftRequest struct {
	InvoiceID              pulid.ID
	TenantInfo             pagination.TenantInfo
	Memo                   *string
	RemittanceInstructions *string
	EmailSubject           *string
	EmailBody              *string
	EmailTo                *[]string
	EmailCC                *[]string
	EmailBCC               *[]string
	AttachmentIDs          *[]pulid.ID
}

type InvoicePreviewRequest struct {
	InvoiceID  pulid.ID
	TenantInfo pagination.TenantInfo
	BaseURL    string
}

type InvoicePreviewResult struct {
	Content     []byte `json:"-"`
	ContentType string `json:"contentType"`
	FileName    string `json:"fileName"`
	SizeBytes   int64  `json:"sizeBytes"`
}

type GenerateInvoicePDFResult struct {
	InvoiceID     pulid.ID `json:"invoiceId"`
	WorkflowID    string   `json:"workflowId"`
	WorkflowRunID string   `json:"workflowRunId"`
	Status        string   `json:"status"`
}

type InvoiceSendPlanRequest struct {
	InvoiceID  pulid.ID
	TenantInfo pagination.TenantInfo
	BaseURL    string
}

type InvoiceSendRequest struct {
	InvoiceID  pulid.ID
	TenantInfo pagination.TenantInfo
	BaseURL    string
}

type AutoSendInvoiceAfterPDFGenerationRequest struct {
	InvoiceID  pulid.ID
	TenantInfo pagination.TenantInfo
	BaseURL    string
}

type InvoiceSendPlan struct {
	InvoiceID            pulid.ID               `json:"invoiceId"`
	ProviderLimitBytes   int64                  `json:"providerLimitBytes"`
	EstimatedBodyBytes   int64                  `json:"estimatedBodyBytes"`
	Parts                []*InvoiceSendPlanPart `json:"parts"`
	Warnings             []string               `json:"warnings"`
	Errors               []string               `json:"errors"`
	Recipients           InvoiceSendRecipients  `json:"recipients"`
	FromEmail            string                 `json:"fromEmail"`
	Headers              map[string]string      `json:"headers"`
	OpenTracking         bool                   `json:"openTracking"`
	Subject              string                 `json:"subject"`
	Body                 string                 `json:"body"`
	InvoicePDFDocumentID pulid.ID               `json:"invoicePdfDocumentId"`
}

type InvoiceSendRecipients struct {
	To  []string `json:"to"`
	CC  []string `json:"cc"`
	BCC []string `json:"bcc"`
}

type InvoiceSendPlanPart struct {
	PartNumber         int                            `json:"partNumber"`
	EstimatedSizeBytes int64                          `json:"estimatedSizeBytes"`
	Attachments        []*InvoiceSendPlanAttachment   `json:"attachments"`
	Links              []*InvoiceSendPlanDocumentLink `json:"links"`
	Warnings           []string                       `json:"warnings"`
}

type InvoiceSendPlanAttachment struct {
	DocumentID   pulid.ID `json:"documentId"`
	FileName     string   `json:"fileName"`
	ContentType  string   `json:"contentType"`
	SizeBytes    int64    `json:"sizeBytes"`
	EncodedBytes int64    `json:"encodedBytes"`
	InvoicePDF   bool     `json:"invoicePdf"`
}

type InvoiceSendPlanDocumentLink struct {
	DocumentID pulid.ID `json:"documentId"`
	FileName   string   `json:"fileName"`
	SizeBytes  int64    `json:"sizeBytes"`
	Reason     string   `json:"reason"`
	URL        string   `json:"url,omitempty"`
}

type InvoiceSendResult struct {
	Invoice  *invoice.Invoice        `json:"invoice"`
	Plan     *InvoiceSendPlan        `json:"plan"`
	Attempts []*invoice.EmailAttempt `json:"attempts"`
}

type DownloadInvoiceDocumentRequest struct {
	Token string `json:"-"`
}

type DownloadInvoiceDocumentResult struct {
	FileName           string `json:"fileName"`
	ContentType        string `json:"contentType"`
	ContentLength      int64  `json:"contentLength"`
	ContentDisposition string `json:"contentDisposition"`
	Body               []byte `json:"-"`
}

type InvoiceService interface {
	List(
		ctx context.Context,
		req *repositories.ListInvoicesRequest,
	) (*pagination.ListResult[*invoice.Invoice], error)
	GetByID(
		ctx context.Context,
		req repositories.GetInvoiceByIDRequest,
	) (*invoice.Invoice, error)
	CreateFromApprovedBillingQueueItem(
		ctx context.Context,
		req *CreateInvoiceFromBillingQueueRequest,
		actor *RequestActor,
	) (*CreateInvoiceFromBillingQueueResult, error)
	CreateFromShipments(
		ctx context.Context,
		req *CreateInvoiceFromShipmentsRequest,
		actor *RequestActor,
	) (*invoice.Invoice, error)
	UpdateDraft(
		ctx context.Context,
		req *UpdateInvoiceDraftRequest,
		actor *RequestActor,
	) (*invoice.Invoice, error)
	RenderPreview(
		ctx context.Context,
		req *InvoicePreviewRequest,
	) (*InvoicePreviewResult, error)
	GeneratePDF(
		ctx context.Context,
		req *InvoicePreviewRequest,
		actor *RequestActor,
	) (*GenerateInvoicePDFResult, error)
	AutoSendInvoiceAfterPDFGeneration(
		ctx context.Context,
		req *AutoSendInvoiceAfterPDFGenerationRequest,
		actor *RequestActor,
	) (*InvoiceSendResult, error)
	PlanSend(
		ctx context.Context,
		req *InvoiceSendPlanRequest,
	) (*InvoiceSendPlan, error)
	Send(
		ctx context.Context,
		req *InvoiceSendRequest,
		actor *RequestActor,
	) (*InvoiceSendResult, error)
	SendFromWorkflow(
		ctx context.Context,
		req *InvoiceSendRequest,
		actor *RequestActor,
	) (*InvoiceSendResult, error)
	ListEmailAttempts(
		ctx context.Context,
		req repositories.ListInvoiceEmailAttemptsRequest,
	) (*pagination.ListResult[*invoice.EmailAttempt], error)
	DownloadSharedDocument(
		ctx context.Context,
		req *DownloadInvoiceDocumentRequest,
	) (*DownloadInvoiceDocumentResult, error)
	Post(
		ctx context.Context,
		req *PostInvoiceRequest,
		actor *RequestActor,
	) (*invoice.Invoice, error)
	EnqueueAutoPost(
		ctx context.Context,
		entity *invoice.Invoice,
		actor *RequestActor,
	) error
}

File: client/apps/web/src/types/invoice.ts
import { z } from "zod";
import { billTypeSchema, billingQueueItemSchema } from "./billing-queue";
import { customerPaymentTermSchema, customerSchema } from "./customer";
import { documentSchema } from "./document";
import { emailMessageSchema } from "./email";
import {
  decimalStringSchema,
  nullableIntegerSchema,
  nullableStringSchema,
  tenantInfoSchema,
} from "./helpers";
import { shipmentSchema } from "./shipment";

export const invoiceStatusSchema = z.enum(["Draft", "Posted"]);
export type InvoiceStatus = z.infer<typeof invoiceStatusSchema>;

export const invoiceSendStatusSchema = z.enum([
  "NotSent",
  "Sending",
  "Sent",
  "PartiallySent",
  "Failed",
]);
export type InvoiceSendStatus = z.infer<typeof invoiceSendStatusSchema>;

export const invoiceLineTypeSchema = z.enum(["Freight", "Accessorial"]);
export type InvoiceLineType = z.infer<typeof invoiceLineTypeSchema>;

export const invoiceLineSchema = z.object({
  ...tenantInfoSchema.shape,
  id: z.string(),
  invoiceId: z.string(),
  lineNumber: z.number().int(),
  type: invoiceLineTypeSchema,
  description: z.string(),
  quantity: decimalStringSchema,
  unitPrice: decimalStringSchema,
  amount: decimalStringSchema,
});
export type InvoiceLine = z.infer<typeof invoiceLineSchema>;

export const settlementStatusSchema = z.enum(["Unpaid", "PartiallyPaid", "Paid"]);
export type SettlementStatus = z.infer<typeof settlementStatusSchema>;

export const invoiceAttachmentSchema = z.object({
  ...tenantInfoSchema.shape,
  id: z.string(),
  invoiceId: z.string(),
  documentId: z.string(),
  selected: z.boolean(),
  sortOrder: z.number().int(),
  createdAt: z.number(),
  updatedAt: z.number(),
  document: documentSchema.nullable().optional(),
});
export type InvoiceAttachment = z.infer<typeof invoiceAttachmentSchema>;

export const invoiceEmailAttemptAttachmentSchema = z.object({
  ...tenantInfoSchema.shape,
  id: z.string(),
  attemptId: z.string(),
  documentId: z.string(),
  fileName: z.string(),
  contentType: z.string(),
  sizeBytes: z.number(),
  encodedBytes: z.number(),
  method: z.enum(["Attached", "Link", "Skipped", "Failed"]),
  shareTokenId: nullableStringSchema,
  reason: nullableStringSchema,
  createdAt: z.number(),
  document: documentSchema.nullable().optional(),
});
export type InvoiceEmailAttemptAttachment = z.infer<typeof invoiceEmailAttemptAttachmentSchema>;

export const invoiceEmailAttemptSchema = z.object({
  ...tenantInfoSchema.shape,
  id: z.string(),
  invoiceId: z.string(),
  emailMessageId: nullableStringSchema,
  attemptNumber: z.number().int(),
  partNumber: z.number().int(),
  totalParts: z.number().int(),
  status: invoiceSendStatusSchema,
  provider: z.string().nullable().optional(),
  providerMessageId: nullableStringSchema,
  toRecipients: z.array(z.string()),
  ccRecipients: z.array(z.string()).nullable().optional(),
  bccRecipients: z.array(z.string()).nullable().optional(),
  subject: z.string(),
  body: nullableStringSchema,
  estimatedSize: z.number(),
  warnings: z.array(z.string()).nullable().optional(),
  error: nullableStringSchema,
  sentAt: nullableIntegerSchema,
  createdById: nullableStringSchema,
  createdAt: z.number(),
  updatedAt: z.number(),
  email: emailMessageSchema.nullish().transform((value) => value ?? undefined),
  attachments: z.array(invoiceEmailAttemptAttachmentSchema).optional().default([]),
});
export type InvoiceEmailAttempt = z.infer<typeof invoiceEmailAttemptSchema>;

export const invoiceSendPlanAttachmentSchema = z.object({
  documentId: z.string(),
  fileName: z.string(),
  contentType: z.string(),
  sizeBytes: z.number(),
  encodedBytes: z.number(),
  invoicePdf: z.boolean(),
});

export const invoiceSendPlanDocumentLinkSchema = z.object({
  documentId: z.string(),
  fileName: z.string(),
  sizeBytes: z.number(),
  reason: z.string(),
  url: z.string().optional(),
});

const stringArraySchema = z
  .array(z.string())
  .nullish()
  .transform((value) => value ?? []);

export const invoiceSendPlanPartSchema = z.object({
  partNumber: z.number().int(),
  estimatedSizeBytes: z.number(),
  attachments: z
    .array(invoiceSendPlanAttachmentSchema)
    .nullish()
    .transform((value) => value ?? []),
  links: z
    .array(invoiceSendPlanDocumentLinkSchema)
    .nullish()
    .transform((value) => value ?? []),
  warnings: stringArraySchema,
});

export const invoiceSendPlanSchema = z.object({
  invoiceId: z.string(),
  providerLimitBytes: z.number(),
  estimatedBodyBytes: z.number(),
  parts: z
    .array(invoiceSendPlanPartSchema)
    .nullish()
    .transform((value) => value ?? []),
  warnings: stringArraySchema,
  errors: stringArraySchema,
  recipients: z.object({
    to: stringArraySchema,
    cc: stringArraySchema,
    bcc: stringArraySchema,
  }),
  fromEmail: z.string().optional().default(""),
  headers: z
    .record(z.string(), z.string())
    .nullish()
    .transform((value) => value ?? {}),
  openTracking: z.boolean().optional().default(false),
  subject: z.string(),
  body: z.string(),
  invoicePdfDocumentId: nullableStringSchema,
});
export type InvoiceSendPlan = z.infer<typeof invoiceSendPlanSchema>;

export const invoiceSendResultSchema = z.object({
  invoice: z.lazy(() => invoiceSchema),
  plan: invoiceSendPlanSchema,
  attempts: z
    .array(invoiceEmailAttemptSchema)
    .nullish()
    .transform((value) => value ?? []),
});
export type InvoiceSendResult = z.infer<typeof invoiceSendResultSchema>;

export const generateInvoicePdfResultSchema = z.object({
  invoiceId: z.string(),
  workflowId: z.string(),
  workflowRunId: z.string(),
  status: z.literal("Queued"),
});
export type GenerateInvoicePdfResult = z.infer<typeof generateInvoicePdfResultSchema>;

export const invoiceSchema = z.object({
  ...tenantInfoSchema.shape,
  id: z.string(),
  billingQueueItemId: z.string(),
  shipmentId: nullableStringSchema,
  orderId: nullableStringSchema,
  customerId: z.string(),
  number: z.string(),
  billType: billTypeSchema,
  status: invoiceStatusSchema,
  paymentTerm: customerPaymentTermSchema,
  currencyCode: z.string(),
  invoiceDate: z.number(),
  dueDate: nullableIntegerSchema,
  postedAt: nullableIntegerSchema,
  shipmentProNumber: z.string().optional().nullable(),
  shipmentBol: z.string().optional().nullable(),
  orderNumber: z.string().optional().nullable(),
  serviceDate: nullableIntegerSchema,
  billToName: z.string(),
  billToCode: nullableStringSchema,
  billToAddressLine1: nullableStringSchema,
  billToAddressLine2: nullableStringSchema,
  billToCity: nullableStringSchema,
  billToState: nullableStringSchema,
  billToPostalCode: nullableStringSchema,
  billToCountry: nullableStringSchema,
  subtotalAmount: decimalStringSchema,
  otherAmount: decimalStringSchema,
  totalAmount: decimalStringSchema,
  appliedAmount: decimalStringSchema,
  settlementStatus: settlementStatusSchema,
  disputeStatus: z.enum(["None", "Disputed"]),
  pdfDocumentId: nullableStringSchema,
  sendStatus: invoiceSendStatusSchema.default("NotSent"),
  sentAt: nullableIntegerSchema,
  sentById: nullableStringSchema,
  lastSendError: nullableStringSchema,
  lastSendWarning: nullableStringSchema,
  memo: nullableStringSchema,
  remittanceInstructions: nullableStringSchema,
  emailSubjectSnapshot: nullableStringSchema,
  emailBodySnapshot: nullableStringSchema,
  emailToSnapshot: z.array(z.string()).nullable().optional(),
  emailCcSnapshot: z.array(z.string()).nullable().optional(),
  emailBccSnapshot: z.array(z.string()).nullable().optional(),
  correctionGroupId: nullableStringSchema,
  supersedesInvoiceId: nullableStringSchema,
  supersededByInvoiceId: nullableStringSchema,
  sourceInvoiceAdjustmentId: nullableStringSchema,
  isAdjustmentArtifact: z.boolean().default(false),
  version: z.number(),
  createdAt: z.number(),
  updatedAt: z.number(),
  billingQueueItem: billingQueueItemSchema.nullish().transform((value) => value ?? undefined),
  shipment: shipmentSchema.nullish().transform((value) => value ?? undefined),
  customer: customerSchema.nullish().transform((value) => value ?? undefined),
  pdfDocument: documentSchema.nullish().transform((value) => value ?? undefined),
  lines: z
    .array(invoiceLineSchema)
    .nullish()
    .transform((value) => value ?? []),
  attachments: z
    .array(invoiceAttachmentSchema)
    .nullish()
    .transform((value) => value ?? []),
  emailAttempts: z
    .array(invoiceEmailAttemptSchema)
    .nullish()
    .transform((value) => value ?? []),
});
export type Invoice = z.infer<typeof invoiceSchema>;

export const updateInvoiceDraftSchema = z.object({
  memo: z.string().optional(),
  remittanceInstructions: z.string().optional(),
  emailSubject: z.string().optional(),
  emailBody: z.string().optional(),
  emailTo: z.array(z.string()).optional(),
  emailCc: z.array(z.string()).optional(),
  emailBcc: z.array(z.string()).optional(),
  attachmentIds: z.array(z.string()).optional(),
});
export type UpdateInvoiceDraft = z.infer<typeof updateInvoiceDraftSchema>;
